Adding `Alpha Features` (#1715)
Azure/azure-dev: cli/azd/docs/featue-toggle.md

> # Feature Toggle

Reviewer: I think this doc is great enough to be thought of as the `product` spec instead of a `proposal` spec. If you agree, let's decide on the proper noun of the feature.

I would vote on:

- Renaming doc to `experimental-features.md`
- Renaming title to `Experimental Features`
- Concisely define the package name as `expfeature` to support the functionality (I think it warrants a package based on what I'm seeing)

The Azure Developer CLI includes a `feature toggle` which allows the cli to ship `in-preview` ( _a.k.a. experimental_ ). This document provides the feature design for it.

## Main requirement

As a customer, I can explicitly ask `azd` to unblock and show any _experimental_ functionality. The features becoming available are not finalized, so they can be changed or removed in future releases.

The strategy to ask `azd` to `toggle` between modes is:

> > azd config set experimental on

Reviewer: I think we also want granular experimental features. This allows users to opt-in to specific features, while not impacting other pieces of the product.

```bash
azd config set experimental on # Turns on all features
azd config set experimental.provision.caching on # Turns on `provision.caching` feature
````

Also, would you recommend instead users did:

`azd config unset experimental`? I think this is better as it avoids `config `pollution. We should treat `off` and `undefined` similarly.

> > azd config set experimental on

Reviewer: Can we please think about supporting `AZD_EXPERIMENTAL_<flag>`? This is a common CLI strategy to allow both. Since we will always emit a warning message indicating experimental features being on, I think this resolves the env pollution problem.

The reason to have this is to allow **trial** of new features, before enabling it for a particular machine.

and

> > azd config set experimental off

Reviewer 2: This gesture also works when the value of `experimental` is an object instead of just "on", which is nice.

## Experimental features

> When toggling experimental mode `on`, the Azure Developer CLI includes the next type of features:

Reviewer: ```suggestion
The Azure Developer CLI offers the following type of features that can be toggled using experimental modes:
```

1. Command or sub-command
    - The command becomes available to be run just like any other command.

>     - Customers are **not** expected to know when a command or sub-command is experimental by just typing the command. This means:
>       - There are not suffix or prefix for a in-preview command.
>       - There is not sub-grouping command for all experimental commands.

Reviewer: Reword for simplicity:

```suggestion
    - The command is designed exactly as how it would be released.
      - There is no suffix or prefix to the command
      - There is no `experimental` command group which experimental commands live under.
```

    - A `warning note` should be displayed then the command is run with a note like:
      > You are running an experimental command. It can be changed or removed in the future.

> 1. Flags for command

Reviewer: ```suggestion
1. New flags for command
```

    - Follows the same rules from `1)`

> 1. Field for `azure.yaml`

Reviewer: ```suggestion
1. New field in `azure.yaml`
```

    - There won't be changes to the `yaml-schema`. This means, setting fields for in-preview features will be syntax highlighted as an error. Customer should expect and ignore those errors.

>     - Similar to 1) and 2), when experimental fields are set within azure.yaml, there will be a warning message displayed while azd loads the `azure.yaml`.

Reviewer: I feel like this may be solved with the general design of "User should see a message like "Running with experimental feature '' enabled." The user already knows he's adding an experimental field, when the syntax error occurs. He just needs to know that `azd` has the experimental feature on.

> 

Reviewer: I think you're missing a group for "All other application features (providers, functionality, behavior)". I also think the general practice (aside from commands), is to do the following:

All features (aside from commands)
1. User should see a message like "Running with experimental feature '<feature>' enabled."

> Note: The Azure Developer CLI will never use the experimental toggling for breaking-changes like:
> - Renaming existing commands or flags.
> - Updating yaml-schema.
> - Changing the existing behavior for any flow.
>
> An experimental feature must be an **add-on only** to the current flows.

## Announcing in-preview features

> As mentioned before, when experimental mode is enabled, the only way a customer is expected to know when one of the experimental features is used, is by getting a warning note after invoking the feature. This means:

Reviewer: Can we rewrite this section in terms of:

Here's how users can get to experimental features:

- Public release notes for any additions or removals
- Changelog notes for precise list

The cli will not expose a tracker list currently.

- The is not way for azd to show a list of experimental features available.

> - User can not individually select which features to activate. All the experiments are enabled from the same switch.

Reviewer 3: I would make the features toggles to be enabled one by one vs an all or nothing.  I could see a use case for enabled all features/experiments but we really do need a per-feature toggle as well as different features will ship with different levels of maturity at different times.

Reviewer 2: I agree here - I think the design Wei outlined (and enabled by the config system) of having the value of `experimental` be an object instead of a string is a good idea.

Author: will do
